tests/unit/s2n_client_renegotiation_info_extension_test.c

```diff
@@ -75,5 +85,108 @@ int main(int argc, char **argv)
+        for (size_t i = 0; i < server_conn->client_hello.cipher_suites.size; i += S2N_TLS_CIPHER_SUITE_LEN) {
+            uint8_t *iana = server_conn->client_hello.cipher_suites.data + i;
+            if (memcmp(iana, renegotiation_info_scsv_iana, S2N_TLS_CIPHER_SUITE_LEN) == 0) {
+                found_renegotiation_info_scsv = true;
+            }
+        }
+        EXPECT_TRUE(found_renegotiation_info_scsv);
+
+        /* Do NOT expect "renegotiation_info" extension */
+        s2n_extension_type_id extension_id = 0;
+        EXPECT_SUCCESS(s2n_extension_supported_iana_value_to_id(
+                s2n_client_renegotiation_info_extension.iana_value, &extension_id));
+        s2n_parsed_extension *extension = &server_conn->client_hello.extensions.parsed_extensions[extension_id];
+        EXPECT_EQUAL(extension->extension.size, 0);
+        EXPECT_EQUAL(extension->extension_type, 0);
+    }
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc5746#3.6
+     *= type=test
+     *# o  The server MUST check if the "renegotiation_info" extension is
+     *# included in the ClientHello.
+     */
+    {
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client_conn);
+
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+
+        /* s2n-tls clients do not send the "renegotiation_info" extension.
+         * Instead, they send the TLS_EMPTY_RENEGOTIATION_INFO_SCSV cipher suite.
+         * See previous test.
+         */
+        EXPECT_SUCCESS(s2n_client_hello_send(client_conn));
+
+        /* Process the extensions on the server */
+        EXPECT_SUCCESS(s2n_stuffer_copy(&client_conn->handshake.io, &server_conn->handshake.io,
+                s2n_stuffer_data_available(&client_conn->handshake.io)));
+        EXPECT_SUCCESS(s2n_parse_client_hello(server_conn));
+        EXPECT_SUCCESS(s2n_extension_list_process(S2N_EXTENSION_LIST_CLIENT_HELLO, server_conn,
+                &server_conn->client_hello.extensions));
+
+        /* Expect secure renegotiation to still be false: no extension received */
+        EXPECT_FALSE(server_conn->secure_renegotiation);
+
+        /* Manually append the "renegotiation_info" extension to the original list. */
+        uint8_t extension[] = {
+                0xff, 0x01, /* extension type: renegotiation_info */
+                0x00, 0x01, /* extension length: 1 */
+                0x00, /* renegotiated_connection length: 0 */
+        };
+        size_t client_hello_size = server_conn->client_hello.raw_message.size;
+        size_t old_extensions_size = server_conn->client_hello.extensions.raw.size;
+        size_t new_extensions_size = old_extensions_size + sizeof(extension);
+        EXPECT_SUCCESS(s2n_stuffer_rewrite(&client_conn->handshake.io));
```

Comment:
I feel like we need a test function that can write custom extensions for us. Not really sure how that would work, but this is something that would be really useful.